### ISSUE TITLE
Feature request: Hooking into esbuild from another build system

### ISSUE THREAD

**Issue by Tatskaari** [evanw/esbuild#817]
Hey guys,

Very impressive stuff you guys have here. Your benchmarks seem very promising and I love that it's written in go, taking full advantage of native performance, and efficient concurrency. I was hoping I could present to you a use case you may not have considered before and see if it's something you'd be willing to consider for your roadmap.

I maintain [please](https://please.build), a high performance, cross language build system also written in go, designed with  massive monorepo in mind. Please follows a similar paradigm to Bazel if you're at all familiar with that. I frequently get asked about how to build Javascript and typescript in Please however the problem is inherently difficult because these paradigms just don't line up. Essentially, there's no sensible level in which Please can hook into these ecosystems. 

**TL;DR: Most bundlers expect to have access to the whole world at all times whereas Please builds are hermetic. A lower level API would go a long way to helping build JS/TS/CSS in Please like build systems.** 

Essentially what I'm looking for is a lower level API from esbuild that I can use to incrementally bundle JS and css files. Conceptually this is similar to `go build` vs. `go tool compile`. I noticed that you're documentation mentions there's an in memory AST for incremental builds. If we can serialise this AST to disk (potentially using [gob](https://golang.org/pkg/encoding/gob/)s), this would allow us to integrate esbuild into Please nicely!

Imagine we have a third party module This might look something like:
```
$ esbuild compile -o node_modules/funky_utils/funky_utils.a node_modules/funky_utilz/funky_utilz.js
$ esbuild compile -o my_module.a entry_point.js # validates the required modules are present on disk in `.a` form i.e.  node_modules/funky_utils/funky_utils.a
$ esbuild link my_module.a -o my_module.bundle.js # reads all the .a files (can be based on their path but Ideally I'd love to pass some sort of module -> .a mapping) and produces a bundled .js file
```

Of course this will be slower than an in-memory dev server but Please has a sophisticated caching system that would massively outweigh any losses here. This approach leaves orchestrating things like invoking the ts compiler, setting up file watcher for dev servers, or generating sources, to Please while leaving esbuild do what it's good at: minifying the code. 

I'm currently trying to get to grips with the esbuild code base but would love to know if this is something you might consider. If this use case is too niche, perhaps we can split out the relevant parts of esbuild into a library and I can write my own CLI around that. It looks like the relevant code lives in the internal package though. 

Let me know what you think :)

**Comment by evanw** [comment]
The fact that there is no API for interacting with the AST is a deliberate design decision. This lets me change the AST format without breaking people using esbuild as a library. I have designed esbuild to be fast enough that you can just run esbuild on source code instead of ASTs. So think of JavaScript source code as a "serialized AST" format.

I think of esbuild as mostly a "linker for the web". It takes care of bundling JavaScript code but how your code ends up as JavaScript is up to you. You can use whatever non-JavaScript technologies you want (e.g. https://www.purescript.org/) as long as it can end up being JavaScript in the end.

The plugin API is not necessary to use esbuild like this. It's just a convenience. It's possible to emulate the plugin API by just transforming JavaScript files ahead of time, writing them to the file system, and then running esbuild once at the end. People just find using esbuild's plugin API and doing this in memory instead of on the file system to be more ergonomic.

So your `compile` step should be any expensive preprocessing you need to do to your source code to turn it into JavaScript (or TypeScript, since esbuild parses both at the same speed). This preprocessing step will likely not use esbuild, or this step may not even be necessary. Then your `link` step is where esbuild comes in. You can completely customize esbuild's path resolution using a plugin if you want some sort of module mapping: https://esbuild.github.io/plugins/#resolve-callbacks. This is equivalent to rewriting the import paths in each JavaScript file, except done on the fly.

> leaving esbuild do what it's good at: minifying the code.

You should be able to use esbuild's `Transform` API to minify code from Go if you're just looking for a minifier.

**Comment by Tatskaari** [comment]
Thanks for the speedy reply and for taking the time to consider my use case. Very much appreciated. I think the major difference between a minifier and what I'm looking for is the idea of linking and compiling against already minified js.

It sounds like esbuild does almost everything I want. I think I understand how I might get "linking" to work. Once I have a bunch of minified js files, I can just run esbuild to join them up into a single .js file. This can work based on the dir structure but if I want to add some smarts to that, I can change the mapping with a plugin. 

The missing part of the puzzle is how much of the work we do upfront and how much is done at linking time. Won't esbuild re-minify the already minified sources including resolving all the `require()`s at runtime? I guess I was imagining something more sophisticated where we can produce a number of already minified .js/.css files, and then when we link, we just merge them together in a relatively inexpensive operation.

One way this could work is by simply concatenating each individually minified file together, and then adding the entrypoint to the bottom of the file. At compile time, we would need to resolve the `requires()` to a variable. Each minified js file might looks like:

```
// foo.js:
const bar = require("bar")

function foo() {
    ...
   bar.bar()
}
// foo.min.js
const foo_<some_fingerprint> = function(){return {foo: function(){return bar_<some_fingerprint>().bar()}}};

// bar.js:
function bar(){
   ...
}
// bar.min.js:
const bar_<some_fingerprint> = ...;

// app.bundle.js:

// foo.min.js here
// bar.min.js here
foo_<some_fingerprint>.foo();
```

This is actually kinda how the webpack DLL plugin works, they're just missing the linking step: 
https://webpack.js.org/plugins/dll-plugin/

Maybe this is something I could achieve with the plugin API? I will try and find time to have a proper look.

**Comment by evanw** [comment]
> Won't esbuild re-minify the already minified sources including resolving all the `require()`s at link-time? 

This may sound strange because minification is usually really slow in other tools, but minification is done during parsing with esbuild so enabling the minifier it's really not a performance hit. In fact sometimes it's even faster than not minifying because it means the code generator generates a smaller output file. Trying to cache minification up front sounds like a premature optimization to me.

**Comment by Tatskaari** [comment]
Ah nice! I think I see the whole picture now. There's a few reasons that I want to minify up front beyond just saving CPU time. Our builds are clustered so the outputs of each step are passed over the network in the worker fleet. Transferring lots of small and bulky files like this is slow. Right now we gzip them but that's slow due to compressing and decompressing each end. I realise this isn't a very universal concern though. I think i can use esbuild to solve this problems:

Step 1 is to create a plugin that will allow me to compile a yarn module into a `module.min.js` file. This plugin should allow resolution of local `requires()` to other files in the same module, but do nothing on `requires()` for other modules. We need some import config that tells esbuild what external modules exist and which don't so that this stage can report any missing third party dependencies. 

Step 2 is for Please to do all our pre-processing as you describe (invoking the typescript compiler etc.) to produce the relevant js files. 

The final step is to use esbuild to link all these together into the final bundle. We need to do this for all the tests as well as the production bundle. This was the point in which I couldn't get webpack to work for me but it sounds like esbuild will be plenty fast enough. 

Thanks so much for all the help. I'll see if I can get started this weekend.

**Comment by Tatskaari** [comment]
Ah! The `--external` flag does exactly what I want. Kudos! This CLI is very intuitive.

**Comment by dror-weiss** [comment]
@Tatskaari , I'm in the process of building a monorepo and want to work with Please. 
How does `--external` helped you, and, more importantly, do you have examples of integrating esbuild with Please?

**Comment by Tatskaari** [comment]
> @Tatskaari , I'm in the process of building a monorepo and want to work with Please. How does `--external` helped you, and, more importantly, do you have examples of integrating esbuild with Please?

It allowed me to convert anything with a valid plugin in es-build into a common module format that we can output from `js_library()`. @tiagovtristao has been working on a plugin here: https://github.com/tiagovtristao/please-js

From what I've heard, it's pretty feature complete but invoking the typescript compiler for each individual target like we do is slow.
